File: src/domain/window.js
/**
 * Run window and UTC daily bucketing (design section 6; flow steps 1 and 3).
 *
 * `run_at` is captured once. The window is the exact half-open UTC range
 * [run_at - 168h, run_at). A rolling 168-hour range normally touches eight UTC calendar
 * dates, so the first and last buckets are usually partial; each bucket therefore carries
 * its own real boundaries and `covered_hours` rather than being assumed to be a full day.
 */

export const WINDOW_HOURS = 168;
const MS_PER_HOUR = 3600000;
const MS_PER_DAY = 24 * MS_PER_HOUR;

/**
 * @typedef {object} DailyBucket
 * @property {string} snapshotDate UTC calendar date, `YYYY-MM-DD`.
 * @property {Date} bucketStart Inclusive.
 * @property {Date} bucketEnd Exclusive.
 * @property {number} coveredHours Real hours of this date inside the window.
 */

/**
 * @typedef {object} RunWindow
 * @property {Date} runAt
 * @property {Date} windowStart Inclusive.
 * @property {Date} windowEnd Exclusive; equals runAt.
 * @property {DailyBucket[]} buckets Ascending by date.
 */

/**
 * UTC calendar date of an instant, as `YYYY-MM-DD`.
 * @param {Date|number|string} at
 * @returns {string}
 */
export function utcDateKey(at) {
  const d = at instanceof Date ? at : new Date(at);
  return d.toISOString().slice(0, 10);
}

/**
 * Midnight UTC starting the calendar date containing `at`.
 * @param {Date|number} at
 * @returns {number} epoch ms
 */
function utcMidnight(at) {
  const ms = at instanceof Date ? at.getTime() : at;
  return Math.floor(ms / MS_PER_DAY) * MS_PER_DAY;
}

/**
 * Build the run window and its daily buckets.
 *
 * @param {Date|string|number} runAt
 * @returns {RunWindow}
 */
export function buildRunWindow(runAt) {
  const end = runAt instanceof Date ? new Date(runAt.getTime()) : new Date(runAt);
  if (Number.isNaN(end.getTime())) throw new TypeError(`invalid run_at: ${String(runAt)}`);
  const start = new Date(end.getTime() - WINDOW_HOURS * MS_PER_HOUR);

  /** @type {DailyBucket[]} */
  const buckets = [];
  // Walk calendar dates from the one containing window_start up to the one containing the
  // last instant inside the window. window_end is exclusive, so a run at exactly midnight
  // must not open an empty bucket for the date it stops at.
  const lastInstant = end.getTime() - 1;
  for (let day = utcMidnight(start); day <= utcMidnight(lastInstant); day += MS_PER_DAY) {
    const bucketStart = Math.max(start.getTime(), day);
    const bucketEnd = Math.min(end.getTime(), day + MS_PER_DAY);
    if (bucketEnd <= bucketStart) continue;
    buckets.push({
      snapshotDate: utcDateKey(day),
      bucketStart: new Date(bucketStart),
      bucketEnd: new Date(bucketEnd),
      coveredHours: (bucketEnd - bucketStart) / MS_PER_HOUR,
    });
  }

  return { runAt: end, windowStart: start, windowEnd: end, buckets };
}

/**
 * Is an instant inside the half-open window?
 * @param {RunWindow} window
 * @param {Date|number} at
 * @returns {boolean}
 */
export function isInWindow(window, at) {
  const ms = at instanceof Date ? at.getTime() : at;
  return ms >= window.windowStart.getTime() && ms < window.windowEnd.getTime();
}

File: src/domain/window.py
"""Run window and UTC daily bucketing (design section 6; flow steps 1 and 3).

``run_at`` is captured once. The window is the exact half-open UTC range
``[run_at - 168h, run_at)``. A rolling 168-hour range normally touches eight UTC calendar
dates, so the first and last buckets are usually partial; each bucket therefore carries its
own real boundaries and ``covered_hours`` rather than being assumed to be a full day.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime, timedelta

__all__ = [
    "WINDOW_HOURS",
    "DailyBucket",
    "RunWindow",
    "build_run_window",
    "utc_date_key",
]

WINDOW_HOURS = 168

_DAY = timedelta(days=1)
_EPOCH = datetime(1970, 1, 1, tzinfo=UTC)


@dataclass(frozen=True, slots=True)
class DailyBucket:
    snapshot_date: str
    """UTC calendar date, ``YYYY-MM-DD``."""
    bucket_start: datetime
    """Inclusive."""
    bucket_end: datetime
    """Exclusive."""
    covered_hours: float
    """Real hours of this date inside the window."""


@dataclass(frozen=True, slots=True)
class RunWindow:
    run_at: datetime
    window_start: datetime
    """Inclusive."""
    window_end: datetime
    """Exclusive; equals ``run_at``."""
    buckets: tuple[DailyBucket, ...]
    """Ascending by date."""

    @property
    def snapshot_dates(self) -> list[str]:
        return [bucket.snapshot_date for bucket in self.buckets]

    def contains(self, at: datetime) -> bool:
        """Is an instant inside the half-open window?"""
        return self.window_start <= _as_utc(at) < self.window_end


def _as_utc(value: datetime) -> datetime:
    """Interpret a naive datetime as UTC; convert an aware one."""
    return value.replace(tzinfo=UTC) if value.tzinfo is None else value.astimezone(UTC)


def utc_date_key(at: datetime) -> str:
    """UTC calendar date of an instant, as ``YYYY-MM-DD``."""
    return _as_utc(at).strftime("%Y-%m-%d")


def _utc_midnight(at: datetime) -> datetime:
    """Midnight UTC starting the calendar date containing ``at``."""
    aware = _as_utc(at)
    return datetime(aware.year, aware.month, aware.day, tzinfo=UTC)


def build_run_window(run_at: datetime) -> RunWindow:
    """Build the run window and its daily buckets."""
    end = _as_utc(run_at)
    start = end - timedelta(hours=WINDOW_HOURS)

    buckets: list[DailyBucket] = []
    # Walk calendar dates from the one containing window_start up to the one containing
    # the last instant inside the window. window_end is exclusive, so a run at exactly
    # midnight must not open an empty bucket for the date it stops at.
    last_instant = end - timedelta(microseconds=1)
    day = _utc_midnight(start)
    final_day = _utc_midnight(last_instant)
    while day <= final_day:
        bucket_start = max(start, day)
        bucket_end = min(end, day + _DAY)
        if bucket_end > bucket_start:
            buckets.append(
                DailyBucket(
                    snapshot_date=utc_date_key(day),
                    bucket_start=bucket_start,
                    bucket_end=bucket_end,
                    covered_hours=(bucket_end - bucket_start).total_seconds() / 3600,
                )
            )
        day += _DAY

    return RunWindow(
        run_at=end,
        window_start=start,
        window_end=end,
        buckets=tuple(buckets),
    )
